File: src/lib/api/git/files.ts
// The repository Files browser — listing, reading, and writing a path — and the
// OS hand-offs (reveal, open, difftool). Mirrors `commands/files.rs`.

import { invoke } from "@/lib/api/invoke";
import type {
  RepoFileContent,
  RepoFileWriteResult,
  RepoFiles,
} from "./types";

export const filesApi = {
  /** Every file in the worktree (tracked + untracked, ignored excluded),
   * repo-relative and sorted — bounded to the backend's listing cap, with
   * `truncated` set when the repository holds more. Whole-repo path search
   * stays with `suggestTreePaths`, which is not limited to this prefix. */
  listRepoFiles: (path: string) => invoke<RepoFiles>("list_repo_files", { path }),

  /** Read one worktree file's text for the read-only viewer. */
  repoFileText: (path: string, file: string, maxBytes?: number) =>
    invoke<RepoFileContent>("repo_file_text", { path, file, maxBytes: maxBytes ?? null }),

  /** The committed (HEAD) text of a file — the baseline for the viewer/editor's
   * uncommitted-change gutter markers. `null` when there's nothing to diff
   * against (unborn HEAD, untracked path, binary/oversized blob). */
  repoFileHeadText: (path: string, file: string) =>
    invoke<string | null>("repo_file_head_text", { path, file }),

  /** Save an edited worktree file (in-app editor, GL-212). The size + opaque
   * state pair identifies the exact target snapshot the draft was based on;
   * Rust refuses same-size edits and atomic replacements too. Resolves with the
   * next lease for sequential saves. Overwrite-only; binary targets/content are
   * refused. */
  writeRepoFile: (path: string, file: string, content: string, expectedSize: number, expectedState: string) =>
    invoke<RepoFileWriteResult>("write_repo_file", {
      path,
      file,
      content,
      expectedSize,
      expectedState,
    }),

  /** Reveal a repo-relative path in Finder / Explorer / the system file manager. */
  revealInFileManager: (path: string, file: string) =>
    invoke<string>("reveal_in_file_manager", { path, file }),

  /** Open a repo-relative worktree leaf with the OS default application. */
  openPathDefault: (path: string, file: string) =>
    invoke<string>("open_path_default", { path, file }),

  /** Open a tracked path in the configured `git difftool` against HEAD. */
  openPathDifftool: (path: string, file: string) =>
    invoke<string>("open_path_difftool", { path, file }),
};

File: src-tauri/src/commands/files.rs
//! The repository Files browser: worktree reads, the in-app editor's guarded write, and open/reveal.

use super::{blocking, CommandError};
use crate::git;
use crate::git::types::{RepoFileContent, RepoFileWriteResult};

/// Every file in the worktree (tracked + untracked, ignored excluded),
/// repo-relative and sorted. The status pass can be expensive on large repos,
/// so run it on the blocking pool like `commit_graph`.
#[tauri::command]
pub async fn list_repo_files(path: String) -> Result<Vec<String>, CommandError> {
    blocking(move || git::status::list_repo_files(&path).map_err(|e| e.to_string())).await
}

/// Read one worktree file's text for the read-only viewer. Binary/oversized
/// content comes back as flags, never raw bytes. Reads up to a couple MiB off
/// disk, so it runs on the blocking pool (like `read_binary_blob`).
#[tauri::command]
pub async fn repo_file_text(
    path: String,
    file: String,
    max_bytes: Option<u64>,
) -> Result<RepoFileContent, CommandError> {
    blocking(move || {
        git::status::repo_file_text(&path, &file, max_bytes).map_err(|e| e.to_string())
    })
    .await
}

/// The committed (HEAD) text of one file — the baseline for the viewer/editor's
/// uncommitted-change gutter markers. `None` when there's nothing to diff
/// against (unborn HEAD, untracked path, binary/oversized blob). Reads a blob,
/// so it runs on the blocking pool like the worktree read.
#[tauri::command]
pub async fn repo_file_head_text(
    path: String,
    file: String,
) -> Result<Option<String>, CommandError> {
    blocking(move || git::status::repo_file_head_text(&path, &file).map_err(|e| e.to_string()))
        .await
}

/// Save an edited worktree file back to disk for the in-app file editor. A
/// guarded, atomic write (overwrite-only, binary + exact-state refusals, temp +
/// rename) that runs on the blocking pool like the read; resolves with the next
/// exact-state lease for sequential saves.
#[tauri::command]
pub async fn write_repo_file(
    path: String,
    file: String,
    content: String,
    expected_size: u64,
    expected_state: String,
) -> Result<RepoFileWriteResult, CommandError> {
    blocking(move || {
        git::write::files::write_repo_file(&path, &file, &content, expected_size, &expected_state)
    })
    .await
}

#[tauri::command]
pub async fn reveal_in_file_manager(path: String, file: String) -> Result<String, CommandError> {
    blocking(move || git::write::reveal::reveal_in_file_manager(&path, &file)).await
}

#[tauri::command]
pub async fn open_path_default(path: String, file: String) -> Result<String, CommandError> {
    blocking(move || git::write::open_path::open_path_default(&path, &file)).await
}

#[tauri::command]
pub async fn open_path_difftool(path: String, file: String) -> Result<String, CommandError> {
    blocking(move || git::write::open_path::open_path_difftool(&path, &file)).await
}
